drivers/ethocan/ethocan.c

```diff
@@ -0,0 +1,604 @@
+    if (new_state < ETHOCAN_STATE_BLOCKED || new_state > ETHOCAN_STATE_SEND) {
+        goto exit;
+    }
+
+    /* Call state specific function */
+    while (1) {
+        uint8_t next_state = new_state;
+        switch (next_state) {
+            case ETHOCAN_STATE_BLOCKED: next_state = state_blocked(ctx, old_state); break;
+            case ETHOCAN_STATE_RECV: next_state = state_recv(ctx, old_state); break;
+            case ETHOCAN_STATE_SEND: next_state = state_send(ctx, old_state); break;
+        }
+
+        if (next_state == new_state) {
+            /* No state change occured within the state's function */
+            break;
+        }
+        else {
+            old_state = new_state;
+            new_state = next_state;
+        }
+    }
+
+    /* Invalid state has been set in state function */
+    if (new_state < ETHOCAN_STATE_BLOCKED || new_state > ETHOCAN_STATE_SEND) {
+        goto exit;
+    }
+
+    ctx->state = new_state;
+    mutex_unlock(&ctx->state_mtx);
+
+exit:
+    irq_restore(irq_state);
+    return ctx->state;
+}
+
+static void _isr_uart(void *arg, uint8_t c)
+{
+    ethocan_t *dev = (ethocan_t *) arg;
+
+    dev->uart_octect = c;
+    state(dev, ETHOCAN_SIGNAL_UART);
+}
+
+static void _isr_gpio(void *arg)
+{
+    ethocan_t *dev = (ethocan_t *) arg;
+
+    state(dev, ETHOCAN_SIGNAL_GPIO);
+}
+
+static void _isr_xtimer(void *arg)
+{
+    ethocan_t *dev = (ethocan_t *) arg;
+
+    state(dev, ETHOCAN_SIGNAL_XTIMER);
+}
+
+static void clear_recv_buf(ethocan_t *ctx)
+{
+    int irq_state = irq_disable();
+
+    ctx->recv_buf_ptr = 0;
+    flag_clear(ctx, ETHOCAN_FLAG_RECV_BUF_DIRTY);
+    flag_clear(ctx, ETHOCAN_FLAG_END_RECEIVED);
+    flag_clear(ctx, ETHOCAN_FLAG_ESC_RECEIVED);
+    irq_restore(irq_state);
+}
+
+static void _isr(netdev_t *netdev)
+{
+    ethocan_t *ctx = (ethocan_t *) netdev;
+    int irq_state, dirty, end;
+
+    /* Get current flags */
+    irq_state = irq_disable();
+    dirty = flag_isset(ctx, ETHOCAN_FLAG_RECV_BUF_DIRTY);
+    end = flag_isset(ctx, ETHOCAN_FLAG_END_RECEIVED);
+    irq_restore(irq_state);
+
+    /* If the recveive buffer does not contain any data just abort ... */
+    if (!dirty) {
+        DEBUG("ethocan _isr(): no frame -> drop\n");
+        return;
+    }
+
+    /* If we havn't received a valid END octect just drop the incomplete frame. */
+    if (!end) {
+        DEBUG("ethocan _isr(): incomplete frame -> drop\n");
+        clear_recv_buf(ctx);
+        return;
+    }
+
+    /* The set dirty flag prevents recv_buf or recv_buf_ptr from being
+     * touched in ISR context. Thus, it is safe to work with them without
+     * IRQs being disabled or mutexes being locked. */
+
+    /* Check the dst mac addr if the iface is not in promiscous mode */
+    if (!opt_isset(ctx, ETHOCAN_OPT_PROMISCUOUS)) {
+        ethernet_hdr_t *hdr = (ethernet_hdr_t *) ctx->recv_buf;
+        if ((hdr->dst[0] & 0x1) == 0 && memcmp(hdr->dst, ctx->mac_addr, ETHERNET_ADDR_LEN) != 0) {
```

Comment:
Is this bit used to indicate broadcast?
Why not check for the broadcast address?

Reply:
This bit identifies multicast and broadcast addresses. If I understood the Ethernet spec correctly, this bit is only zero if it is a unicast frame.

Reply:
Ah, I did not know that!